src/app/clusters/energy-evse-server/energy-evse-server.cpp

```diff
@@ -0,0 +1,438 @@
+        if ((!HasFeature(Feature::kChargingPreferences)) ||
+            (!SupportsOptAttr(OptionalAttributes::kSupportsApproximateEvEfficiency)))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        uint16_t newValue;
+        ReturnErrorOnFailure(aDecoder.Decode(newValue));
+        ReturnErrorOnFailure(mDelegate.SetApproximateEVEfficiency(newValue));
+        return CHIP_NO_ERROR;
+    }
+
+    default:
+        // Unknown attribute; return error.  None of the other attributes for
+        // this cluster are writable, so should not be ending up in this code to
+        // start with.
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+    }
+}
+
+// CommandHandlerInterface
+
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context)
+{
+    using namespace Commands;
+
+    for (auto && id : {
+             Disable::Id,
+             EnableCharging::Id,
+         })
+    {
+        if (callback(id, context) == Loop::Break)
+        {
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    if (HasFeature(Feature::kV2x))
+    {
+        callback(EnableDischarging::Id, context);
+    }
+    if (HasFeature(Feature::kChargingPreferences))
+    {
+        callback(SetTargets::Id, context);
+        callback(GetTargets::Id, context);
+        callback(ClearTargets::Id, context);
+    }
+    if (SupportsOptCmd(OptionalCommands::kSupportsStartDiagnostics))
+    {
+        callback(StartDiagnostics::Id, context);
+    }
+    return CHIP_NO_ERROR;
+}
+
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    using namespace Commands;
+
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case Disable::Id:
+        HandleCommand<Disable::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleDisable(ctx, commandData); });
+        return;
+    case EnableCharging::Id:
+        HandleCommand<EnableCharging::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleEnableCharging(ctx, commandData); });
+        return;
+    case EnableDischarging::Id:
+        if (!HasFeature(Feature::kV2x))
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath, Status::UnsupportedCommand);
+        }
+        else
+        {
+            HandleCommand<EnableDischarging::DecodableType>(handlerContext, [this](HandlerContext & ctx, const auto & commandData) {
+                HandleEnableDischarging(ctx, commandData);
+            });
+        }
+        return;
+    case StartDiagnostics::Id:
+        HandleCommand<StartDiagnostics::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleStartDiagnostics(ctx, commandData); });
+        return;
+    case SetTargets::Id:
+        if (!HasFeature(Feature::kChargingPreferences))
```

Comment:
Note: This does not address. https://github.com/project-chip/connectedhomeip/pull/30857#discussion_r1420974095.  Please do file that followup issue I asked for.

Reply:
Is using a template the only way of resolving this?

Reply:
Ah - I missed the if (!SupportsOptCmd(OptionalCommands::kSupportsStartDiagnostics)) in the InvokeCommands() function, but had it in EnumerateCommands.

Reply:
@jamesharrow Using a template might be the only way of doing it.  It's hard to say.  Again: please just file a followup issue, link it here and in the code.

Reply:
Raised https://github.com/project-chip/connectedhomeip/issues/30994  to track this